README.md
# kube cookbook

The kube cookbook is a Library cookbook that provides custom resources for
managing various components of a Kubernetes cluster.

## Requirements

### Chef

* Chef Infra client 15 or higher

### Platforms

* Ubuntu 16.04+
* Debian 8+
* RHEL 7+

### Cookbook Dependencies

- None

## Usage

* Add `'kube', '~> 5.0'` to your cookbook's `metadata.rb`.
* Use the resources shipped in this cookbook in your recipes the same way you
  use core Chef resources like file, template, directory, package, etc.

```ruby
# Master
kube_apiserver 'default' do
  service_cluster_ip_range '10.0.0.1/24'
  etcd_servers 'http://127.0.0.1:2379'
  insecure_bind_address '0.0.0.0' # for convenience
  action %w(create start)
end

kube_scheduler 'default' do
  master '127.0.0.1:8080'
  action %w(create start)
end

kube_controller_manager 'default' do
  master '127.0.0.1:8080'
  action %w(create start)
end

# Node

kubelet_service 'default' do
  api_servers 'http://127.0.0.1:8080'
  config '/etc/kubernetes/manifests'
  cluster_dns '10.0.0.10'
  cluster_domain 'cluster.local'
  action %w(create start)
end

kube_proxy 'default' do
  master '127.0.0.1:8080'
  action %w(create start)
end
```

The test cookbook ran under test-kitchen provide good usage examples.  It is
found in `test/cookbooks/kube_test`.

## Resources Overview

Components for a Kubernetes node:

* `kubelet`
* `kube_proxy`

Components for a Kubernetes master:

* `kube_apiserver`
* `kube_scheduler`
* `kube_controller_manager`

### Common Properties

All the above resources will contain the following properties:

* `remote` - The URL of where a corresponding component's binary can be
  downloaded.  The default value points to the official Kubernetes release URL of
  each component.  Check each resource for the default value of each component.
* `version` - The version of the Kubernetes artifact to pull down.  Defaults to
  `'1.7.5'`.  NOTE: This will be ignored if you set the `remote` property
  instead.
* `checksum` - The SHA256 hash of the Kubernetes component's binary.
* `run_user` - The user in which to run the Kubernetes user.  Defaults to
  `'kubernetes'`.
* `file_ulimit` - The file ulimit value to set for the services - Integer.
  Defaults to `65536`

### Common Actions

All the above resources will contain the following actions:

* `create` - Download the Kubernetes component's binary to `/usr/bin`.
* `start` - Starts the Kubernetes component managed through a systemd unit.

## Resource Properties

Each resource' set of unique properties corresponds to the options in the
Kubernetes component they represent:

* `kube_apiserver` - <https://github.com/kubernetes/kubernetes.github.io/blob/release-1.7/docs/admin/kube-apiserver.md>
* `kube_controller_manager` - <https://github.com/kubernetes/kubernetes.github.io/blob/release-1.7/docs/admin/kube-controller-manager.md>
* `kube_scheduler` - <https://github.com/kubernetes/kubernetes.github.io/blob/release-1.7/docs/admin/kube-scheduler.md>
* `kubelet` - <https://github.com/kubernetes/kubernetes.github.io/blob/release-1.7/docs/admin/kubelet.md>
* `kube_proxy` - <https://github.com/kubernetes/kubernetes.github.io/blob/release-1.7/docs/admin/kube-proxy.md>

In general, a command line flag of the form `--long-option` will correspond to a
custom resource property called `long_option`.

### Extending Properties

When newer versions of Kubernetes are released, components might introduce and
deprecate some commandline flags that are not yet hard-coded as properties.

To add these properties, a wrapper cookbook can be written like the following:

```ruby
# wrapper-cookbook/metadata.rb
name 'wrapper-cookbook'
depends 'kube', '~> 5.0' # Make sure you have this

# wrapper-cookbook/libraries/apiserver.rb
class KubernetesCookbook::KubeApiserver
  property :something_only_in_kubernetes9000
end
```

The `kube_apiserver` resource can now use the new commandline flag available in
Kubernetes v9000 like the following:

```ruby
# wrapper-cookbook/recipes/default.rb
kube_apiserver 'default' do
  something_only_in_kubernetes9000 'someflag'
end
```

## License

Copyright 2016-2017 Allan Espinosa

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# 5.0.0

* Support Chef Infra Client 16+
* Drop Chef Infra Client 12-13 support

# 4.0.0

* Add kubernetes 1.12 options
* Remove default value for kubeconfig  
* Add --rotate-certificates in kubelet

# 3.0.0

* Use kubernetes 1.7.6 by default.
* Add the `:version` property to pull version binaries upstream.
* Support other systemd-based platforms like CentOS and Ubuntu.
* Bump the minimum version required for the chef-client.
* Document how to anticipate newer versions of Kubernetes.

# 2.0.3

* Bump more metadata to pass Supermarket quality metrics.

# 2.0.2

* Bump documentation and other metadata information to pass Supermarket quality
  metrics.

# 2.0.1

* Document how to use the custom resources
* Update unit tests for Chef 13

# 2.0.0

* Update resource properties to commandline flags in 1.4.0

# 1.1.0

* Update resource properties to reflect kube 1.3.x commandline flags (#13)

# 1.0.0

* Bump kubernetes binaries to 1.3.6
* Remove cluster-name to the api-server parameters (#12)

# 0.4.0

* Set systemd unit flags to automatically start on boot
* Boot the kubelet only after the container runtime is available

# 0.3.0

* Introduce `remote` and `checksum` properties for downloading the binaries
* Fix hard-wired kube-proxy command in its systemd unit

# 0.2.0

* Complete commandline flags as resource properties for all the custom resources:

# 0.1.1

* Specify the cookbook in inline templates
* Move recipe to a test cookbook fixture

# 0.1.0

* Bare resources to boot a cluster


libraries/kube_apiserver.rb
# Copyright 2016-2017 Allan Espinosa
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module KubernetesCookbook
  # Resource to manage a Kubernetes API server
  class KubeApiserver < Chef::Resource
    resource_name :kube_apiserver
    provides :kube_apiserver

    property :version, String, default: '1.7.6'
    property :remote, String,
      default: lazy { |r|
        'https://storage.googleapis.com/kubernetes-release' \
        "/release/v#{r.version}/bin/linux/amd64/kube-apiserver"
      }
    property :checksum, String,
      default: '55426dc77997839ea67658ab08e5023de31b432a6a5df5d41c4f51ef0fc8485c'
    property :run_user, String, default: 'kubernetes'
    property :file_ulimit, Integer, default: 65536

    action :create do
      remote_file "kube-apiserver binary version: #{new_resource.version}" do
        path apiserver_path
        mode '0755'
        source new_resource.remote
        checksum new_resource.checksum
      end
    end

    action :start do
      user 'kubernetes' do
        action :create
        only_if { new_resource.run_user == 'kubernetes' }
      end

      directory '/var/run/kubernetes' do
        owner new_resource.run_user
      end

      template '/etc/tmpfiles.d/kubernetes.conf' do
        source 'systemd/tmpfiles.erb'
        cookbook 'kube'
      end

      systemd_contents = {
        Unit: {
          Description: 'Kubernetes API Server',
          Documentation: 'https://k8s.io',
          After: 'network.target',
        },
        Service: {
          Type: 'notify',
          User: new_resource.run_user,
          ExecStart: generator.generate,
          Restart: 'on-failure',
          LimitNOFILE: new_resource.file_ulimit,
        },
        Install: {
          WantedBy: 'multi-user.target',
        },
      }

      systemd_unit 'kube-apiserver.service' do
        content(systemd_contents)
        action :create
        notifies :restart, 'service[kube-apiserver]', :immediately
      end

      service 'kube-apiserver' do
        action %w(enable start)
      end
    end

    def generator
      CommandGenerator.new apiserver_path, self
    end

    def apiserver_path
      '/usr/sbin/kube-apiserver'
    end

    private

    def file_cache_path
      Chef::Config[:file_cache_path]
    end
  end

  # Commandline-related properties
  # Reference: https://kubernetes.io/docs/admin/kube-apiserver/
  class KubeApiserver
    property :admission_control, default: 'AlwaysAdmit'
    property :admission_control_config_file
    property :advertise_address
    property :allow_privileged
    property :anonymous_auth, default: true
    property :apiserver_count, default: 1
    property :audit_log_maxage
    property :audit_log_maxbackup
    property :audit_log_maxsize
    property :audit_log_path
    property :audit_policy_file
    property :audit_webhook_config_file
    property :audit_webhook_mode, default: 'batch'
    property :authentication_token_webhook_cache_ttl, default: '2m0s'
    property :authentication_token_webhook_config_file
    property :authorization_mode, default: 'AlwaysAllow'
    property :authorization_policy_file
    property :authorization_webhook_cache_authorized_ttl, default: '5m0s'
    property :authorization_webhook_cache_unauthorized_ttl, default: '30s'
    property :authorization_webhook_config_file
    property :azure_container_registry_config
    property :basic_auth_file
    property :bind_address, default: '0.0.0.0'
    property :cert_dir, default: '/var/run/kubernetes'
    property :client_ca_file
    property :cloud_config
    property :cloud_provider
    property :contention_profiling
    property :cors_allowed_origins, default: []
    property :delete_collection_workers, default: 1
    property :deserialization_cache_size, default: 50_000
    property :enable_admission_plugins
    property :enable_aggregator_routing
    property :enable_garbage_collector, default: true
    property :enable_logs_handler, default: true
    property :enable_swagger_ui
    property :etcd_cafile
    property :etcd_certfile
    property :etcd_keyfile
    property :etcd_prefix, default: '/registry'
    property :etcd_quorum_read
    property :etcd_servers, default: []
    property :etcd_servers_overrides, default: []
    property :event_ttl, default: '1h0m0s'
    property :experimental_bootstrap_token_auth
    property :experimental_encryption_provider_config
    property :experimental_keystone_ca_file
    property :experimental_keystone_url
    property :external_hostname
    property :feature_gates
    property :google_json_key
    property :insecure_allow_any_token
    property :insecure_bind_address, default: '127.0.0.1'
    property :insecure_port, default: 8080
    property :kubelet_certificate_authority
    property :kubelet_client_certificate
    property :kubelet_client_key
    property :kubelet_https, default: true
    property :kubelet_preferred_address_type, default: %w(Hostname InternalDNS InternalIP ExternalDNS ExternalIP)
    property :kubelet_port, default: 10_255
    property :kubelet_timeout, default: '5s'
    property :kubernetes_service_node_port
    property :master_service_namespace, default: 'default'
    property :max_connection_bytes_per_sec, default: 0
    property :max_mutating_requests_inflight, default: 200
    property :max_requests_inflight, default: 400
    property :min_request_timeout, default: 1800
    property :oidc_ca_file
    property :oidc_client_id
    property :oidc_groups_claim
    property :oidc_issuer_url
    property :oidc_username_claim, default: 'sub'
    property :profiling, default: true
    property :proxy_client_cert_file
    property :proxy_client_key_file
    property :repair_malformed_updates, default: true
    property :requestheader_allowed_names
    property :requestheader_client_ca_file
    property :requestheader_extra_headers_prefix
    property :requestheader_group_headers
    property :requestheader_username_headers
    property :runtime_config
    property :secure_port, default: 6443
    property :service_account_key_file
    property :service_account_lookup
    property :service_cluster_ip_range
    property :service_node_port_range, default: '30000-32767'
    property :ssh_keyfile
    property :ssh_user
    property :storage_backend
    property :storage_media_type, default: 'application/json'
    property :storage_versions
    property :target_ram_mb
    property :tls_ca_file
    property :tls_cert_file
    property :tls_private_key_file
    property :tls_sni_cert_key
    property :tls_private_key_file
    property :token_auth_file
    property :watch_cache, default: true
    property :watch_cache_sizes, default: []

    property :v, default: 0 # TODO: move to common class
  end
end
